tests/subsys/compress/decompression/lzma/src/main.c

```diff
@@ -0,0 +1,559 @@
+	zassert_not_equal(implementation->deinit, NULL,
+			  "Expected deinit element to not be null\n");
+	zassert_not_equal(implementation->reset, NULL, "Expected reset element to not be null\n");
+	zassert_equal(implementation->compress, NULL, "Expected compress element to be null\n");
+	zassert_not_equal(implementation->decompress_bytes_needed, NULL,
+			  "Expected decompress_bytes_needed element to not be null\n");
+	zassert_not_equal(implementation->decompress, NULL,
+			  "Expected decompress to not be null\n");
+}
+
+ZTEST(nrf_compress_decompression, test_invalid_implementation)
+{
+	struct nrf_compress_implementation *implementation = NULL;
+
+	implementation = nrf_compress_implementation_find(NRF_COMPRESS_TYPE_COUNT);
+
+	zassert_equal(implementation, NULL, "Expected implementation to be null\n");
+
+	implementation = nrf_compress_implementation_find(NRF_COMPRESS_TYPE_COUNT + 300);
+
+	zassert_equal(implementation, NULL, "Expected implementation to be null\n");
+
+	implementation = nrf_compress_implementation_find(((uint16_t)-1));
+
+	zassert_equal(implementation, NULL, "Expected implementation to be null\n");
+}
+
+ZTEST(nrf_compress_decompression, test_valid_data_decompression)
+{
+	int rc;
+	uint32_t pos;
+	uint32_t offset;
+	uint8_t *output;
+	uint32_t output_size;
+	uint32_t total_output_size = 0;
+	uint8_t output_sha[SHA256_SIZE] = { 0 };
+	struct nrf_compress_implementation *implementation;
+	mbedtls_sha256_context ctx;
+
+	mbedtls_sha256_init(&ctx);
+        rc = mbedtls_sha256_starts(&ctx, false);
+	zassert_ok(rc, "Expected mbedtls sha256 start to be successful\n");
+
+	implementation = nrf_compress_implementation_find(NRF_COMPRESS_TYPE_LZMA);
+
+	pos = 0;
+
+	rc = implementation->init();
+	zassert_ok(rc, "Expected init to be successful\n");
+
+	rc = implementation->decompress_bytes_needed();
+	zassert_equal(rc, 2, "Expected to need 2 bytes for LZMA header\n");
+
+	rc = implementation->decompress(&dummy_data_input[pos], rc, false, &offset, &output,
+					&output_size);
+	zassert_ok(rc, "Expected header decompress to be successful\n");
+	pos += offset;
+
+	while (pos < sizeof(dummy_data_input)) {
+		rc = implementation->decompress_bytes_needed();
+		zassert_equal(rc, CONFIG_NRF_COMPRESS_CHUNK_SIZE,
+			      "Expected to need chunk size bytes for LZMA data\n");
+
+		if ((pos + rc) >= sizeof(dummy_data_input)) {
+			rc = implementation->decompress(&dummy_data_input[pos],
+							(sizeof(dummy_data_input) - pos), true,
+							&offset, &output, &output_size);
+			pos += 1;
+		} else {
+			rc = implementation->decompress(&dummy_data_input[pos], rc, false, &offset,
+							&output, &output_size);
+		}
+
+		zassert_ok(rc, "Expected data decompress to be successful\n");
+
+		total_output_size += output_size;
+
+		if (output_size > 0) {
+			rc = mbedtls_sha256_update(&ctx, output, output_size);
+			zassert_ok(rc, "Expected hash update to be successful\n");
+		}
+
+		pos += offset;
```

Comment:
offset is how much of the input data to the compression function has been used, you might feed in e.g. 256 bytes but not all of it might be used and you might need to read in 256 bytes for the next time from a different offset, so it might be the chunk size but it might not be (no guarantee), depending upon the compression system used and the current state of decompression